### DISCUSSION TITLE
QT in Docker with Web

### DISCUSSION THREAD

**Discussion by eskimos20** [cagnulein/qdomyos-zwift#2948] [Q&A] [answered]
Hi! 

I installed docker on a ubuntu server. I ran the docker file. Everything is up and running and I can access the webserver after I did some config so i can access it from http:/ip:8080

But... what exactly should I see when trying to access it... for me I get a blank page with:

QtWebGL - Waiting for connection..

Do I need to start something else.. I thought I would get to some config page and be able to run QT app from docker... what am I doing wrong this time ? 🗡️ 

/Eskimos

**Comment by cagnulein** [comment]
@nix155 could you help?

**Reply by nix155** [reply]
@cagnulein I'll try ))
@eskimos20 Could you provide a bit more detail—what specific file was launched, whether your host has Bluetooth, and so on?

**Reply by eskimos20** [reply]
I have a small Lenovo ThinkCentre .. a Intel 5 Core I5 with built in wifi, bluetooth 5.1 and so on. 

I used this dockerfile: https://github.com/cagnulein/qdomyos-zwift/blob/master/docker/linux_webgl/Dockerfile

Seems updated and it worked well with install. Just needed to run it with:

sudo docker run -d -p 8080:8080 myimage

I can access both the container and the web interface but I get a blank page... just saying waiting for connection. 

I think the problem might be that It cant access my Lenovos bluetooth.. or cant find it. 

At least my bike is not connected since it should be a blue light.. but I can access the Wahoo kickr 0000 33 option from Zwift.. so it might be that it has started the program but it wont connect to the trainer. Maybe bluetooth is not exposed some how.

**Reply by cagnulein** [reply]
From the docker can we have the log?

**Reply by eskimos20** [reply]
Yes.. what is the location in the docker image to the log file ?

**Comment by nix155** [comment]
In both containers,  
**linux_gui_vnc**  
**linux_webgl**,  
a two-stage build process is implemented. First, everything is compiled in the build container, and then the binary and required libraries are copied to the runtime container, which is then launched.  

If you encounter any issues, I would recommend taking the **linux_webgl** directory, building the image from it (using `build.sh`), and running it (`docker compose up -d`) without the GUI by uncommenting the following line in `docker-compose`:  
```yaml
command: ["qdomyos-zwift", "-no-gui"]
```
After that, you can inspect the container logs:  
```bash
docker logs qdomyos-zwift-webgl
```

**Comment by nix155** [comment]
For both directories, the process is the same: copy everything from the **linux_webgl** directory into a single directory, make the `build.sh` script executable (use `chmod +x build.sh`), and then build the image using `build.sh`. After the build, launch the image from the same directory (ensuring you have the `docker-compose.yml` file) with the command:  
```bash
docker compose up -d
```

**Reply by eskimos20** [reply]
I tried to run just the QT that was in bin and found this:

Fri Dec 27 20:16:06 2024 1735330566764 Warning: Cannot find a running Bluez. Please check the Bluez installation.
Cannot find a running Bluez. Please check the Bluez installation.
Fri Dec 27 20:16:06 2024 1735330566764 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString&) "no bluetooth dongle found!"
"no bluetooth dongle found!"

So it is not finding bluetooth :D

**Comment by nix155** [comment]
There might be issues with the GUI in both versions (although there shouldn’t be any on Ubuntu), but the GUI isn’t critical—everything can be configured via a config file, and you can view the logs in the console.  

In any case, first, ensure everything works correctly without the GUI, and then add the GUI later if desired. I tested it on a laptop with Ubuntu/Fedora, and it worked.

**Comment by nix155** [comment]
Bluetooth not found ))  
I’m not sure why you’d want a server on a laptop—just set up a regular workstation.

**Reply by eskimos20** [reply]
Its not a laptop... its a thinkcentre... its like a small Intel NUC. 

I have some options.

1. Run it with gui (it will work) like a regular workstation
2. Run it without gui but I have not figured out how to set the config file, and what parameters should be used to add zwift click.

Are there any config examples ?

**Reply by nix155** [reply]
Yes, of course, and I mentioned this above.  
Do you have experience working with Docker and Docker Compose?

**Reply by eskimos20** [reply]
I have only a little knowledge about docker so I will have to read and google some :D

**Reply by cagnulein** [reply]
@eskimos20 chatgpt for these tasks is perfect (claude.ai even better, it's the one that i'm using since 6 month)

**Reply by nix155** [reply]
All the settings for building the image include the Dockerfile, a build script (hehe) called `build.sh`, and for running, a `docker-compose.yml` file. The command to start it is `docker compose up -d` (from the directory containing the compose file).  
For learning, I would recommend YouTube—it's a great resource and can be combined with a practice environment—and this is not a joke.

**Comment by eskimos20** [comment]
Hi! Ok now i got it to work.

An spinning green circle and then connection to my trainer.

Thanks 😊

Den lör 28 dec. 2024 12:02nix155 ***@***.***> skrev:

**Comment by nix155** [comment, marked as answer]
I’ll write this down so I won’t forget later: "How to build and run the project from Docker":

1. Copy one of these directories with files:  
   - `linux_gui_vnc`  
   - `linux_webgl`  

   To build the project, you only need the contents of one of these directories, so there’s no need to pull the repository.

2. Navigate to the copied directory:  
   ```bash
   cd directory_name
   ```

3. Grant execution permissions to the build script:  
   ```bash
   chmod +x build.sh
   ```

4. Run the script:  
   ```bash
   ./build.sh
   ```

   After the build, you will have an image (you can check the list using `docker image ls`).

5. To run the project, execute the following command from the directory containing the `docker-compose.yml` file:  
   ```bash
   docker compose up -d
   ```

In case of any errors during the build (or while running), it’s usually enough to update the list of libraries—ChatGPT can help with that.

The first stage (`build`) can be used as an example for building on Linux (e.g., in a GitHub pipeline).

**Comment by nix155** [comment]
Added images to Docker Hub, you can download them with the following commands:  
```bash
docker pull r155/qdomyos-zwift-gui-vnc
docker pull r155/qdomyos-zwift-webgl
```

To run them, you can use the corresponding `docker-compose.yml` files from the `docker/...` repository, updating the image name as follows:  
```yaml
image: r155/qdomyos-zwift-gui-vnc
image: r155/qdomyos-zwift-webgl
```

For local use, you can extract the binary from there. 😊  

When the main repository is updated, the images should also be updated (though this is not guaranteed – it needs to be checked).

**Comment by cagnulein** [comment]
Wow that's amazing, thanks!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mar 31 dic 2024 alle 09:49 nix155 ***@***.***> ha
scritto:

**Comment by nix155** [comment]
Moved from Docker Hub to GitHub, added a build for the Linux binary (with a 1-hour check interval for changes in the source repository).

**Linux Binary**  
[https://github.com/nix155/sync-repo/releases](https://github.com/nix155/sync-repo/releases)

**Docker Images**  
[https://github.com/nix155?tab=packages](https://github.com/nix155?tab=packages)

**DockerHub Actions Code**  
[https://github.com/nix155/sync-repo/tree/main/.github/workflows](https://github.com/nix155/sync-repo/tree/main/.github/workflows)

**Reply by eskimos20** [reply]
I did docker pull r155/qdomyos-zwift-webgl 
Changed the image to r155/qdomyos-zwift-webgl in .yml file
Changed values in .yml to: ["qdomyos-zwift", "-no-gui", "-zwift_play_emulator", "-no-log"]
Edited the restart parameter to "always"
Did a docker compose up -d

All good and working with zwift click! :)

Great work!

**Reply by nix155** [reply]
@eskimos20 - awesome!  
Use the images from [https://github.com/nix155?tab=packages](https://github.com/nix155?tab=packages).  

For example:  
```bash
docker pull ghcr.io/nix155/qdomyos-zwift-webgl:20250102-1413
```  
or  
```bash
docker pull ghcr.io/nix155/qdomyos-zwift-webgl:latest
```  

I moved the image builds to GitHub, made them automatic upon updates to the main repository, and added date-based tags.

**Comment by eskimos20** [comment]
Works with docker pull ghcr.io/nix155/qdomyos-zwift-webgl:latest

1337 :D

**Comment by Vendo233** [comment]
I would like to try it.

what would be the docker-compose script ( for Portainer )

I assume one has to enable BT device in the config.



```
version: '3'
   services:
     qdomyos:
       container_name: qdomyos
       image: ghcr.io/nix155/qdomyos-zwift-gui-vnc:20250201-0914
       volumes:
         - /home/server/qdomyos:**/?**
       devices:
         - "/dev/bus/usb/<bus>/<device>:/dev/bus/usb/<bus>/<device>" 
       restart: always
       network_mode: host
```





my USB Bluetooth dongle

Bus 001 Device 004: ID 2550:8761 Realtek Bluetooth Radio

**Comment by EPiSKiNG** [comment]
Been playing around with this a bit today, and using the docker-compose.yaml and changed the image to r155/qdomyos-zwift-webgl and everything seems to work when I access the web portal, but the UI within the web portal seems to refresh itself every second or two, and blinks the entire screen which is quite annoying.  :)

Any fixes for this?

Thanks!
